Update replace_parameter.sh such that it can accept a filename argument

AndrewLevin opened the pull request:
The current implementation of replace_parameter.sh will replace a parameter in all VFAT config files with a single value. My pull request replaces the parameter based on the contents of a text file, which can have different values for each VFAT.

## Description
When `-f` is passed to the script, the first argument is interpreted as a filename. The filename is expected to contain a list of vfatN value pairs, and the script will replace the value in the config file for vfatN for each pair.

The existing functionality of the script is maintained (however, I found that the current sed syntax incorrectly matches both the `HYST` and `EN_HYST` lines, so I fixed that).

Usage:
```
./replace_parameter.sh <REGISTER> <VALUE> <LINK>
./replace_parameter.sh -f <FILENAME> <REGISTER> <LINK>

        REGISTER - register to be updated dropping the "CFG_" substring

        VALUE - value in decimal to be assigned to REGISTER

        FILENAME - name of a file containing a list of vfat number/value pairs

        Examples:

                ./replace_parameter.sh PULSE_STRETCH 4 1
                ./replace_parameter.sh -f /path/to/NominalDacValues.txt PULSE_STRETCH 1
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
anaDACScan.py (https://github.com/cms-gem-daq-project/gem-plotting-tools/blob/develop/anaDACScan.py) outputs a text file of vfatN value pairs which need to be set on the ctp7. 

## How Has This Been Tested?
Yes, I have tested the script on eagle64 to replace the `HYST` parameter for OH 0.

### Screenshots (if appropriate):

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- Template thanks to https://www.talater.com/open-source-templates/#/page/99 -->

bdorney commented on scripts/replace_parameter.sh:
```diff
@@ -1,27 +1,55 @@
+if (($ISFILE == 1)); then
+    if [ -z ${4+x} ]
+    then
+	    usage
```
Probably some statement here is needed about *which* usage example is expected.  You're using the same help menu for each output.  I think an `echo` after the call of usage to specify the file input case is sufficient.

Same comment for the below.

AndrewLevin replied:
done

jsturdy commented on scripts/replace_parameter.sh:
```diff
@@ -1,27 +1,59 @@
-LINK=$3
+ISFILE=0;
+OPTIND=1
+while getopts "fwhd" opts
```
It seems you have some undocumented/unused flags here (`wd`)

AndrewLevin replied:
fixed

bdorney commented:
What is the status of this PR?

AndrewLevin commented:
@jsturdy, so are you proposing that there be two options `-f` for file input and `-v` for value input, and that it should be required to set exactly one of them?

jsturdy commented:
> so are you proposing that there be two options `-f` for file input and `-v` for value input, and that it should be required to set exactly one of them?

@AndrewLevin, no, rather, `-f` is optional, *but* if it is provided, then it should immediately precede the file path, and the option parser will automatically handle this

AndrewLevin commented:
@jsturdy, `-f` is not really just an optional argument, but it is a switch between two modes of the script: filename input and value input. It is not clear that you understood that. I can change to making what follows `-f` be interpreted as the filename, but then I would also need to change the logic that checks the number of mandatory arguments and interprets them.

jsturdy commented:
When using the `getopts` library, you have several options.
* You can make everything you pass in be associated to a flag/switch.
* You can add to the above positional arguments, but you should have all options/flags/switches first in the call
* You can mix and match positional arguments and options, but this requires special trickery to make sure that all required positional arguments are provided, and all options are parsed

I'm suggesting that you follow the second.  In this case, the usage could be:
```bash
./replace_parameter.sh -f <FILENAME> <REGISTER> <LINK> #valid
./replace_parameter.sh <REGISTER> <LINK> <VALUE> #valid
./replace_parameter.sh <REGISTER> <LINK> -f <FILENAME> #should be invalid, but won't be caught, will treat '-f' as <VALUE>
./replace_parameter.sh <REGISTER> <LINK> <VALUE> <FILENAME> #valid, <FILENAME> would just be ignored
```

If you follow the usage of the `getopts` library, **no** change in your logic is necessary (beyond removing any special treatment you added for the `-f` switch).
You can probably search around and implement the third method, which will make the third example above work correctly and make things overall more robust against non-strict adherence to the `usage`, but it may be time consuming...

AndrewLevin commented:
@jsturdy, I have implemented the change you requested

jsturdy commented:
@AndrewLevin, sorry, I failed to realize before that the positional argument that is the odd one out is the middle one (`<VALUE>`).  This makes the current design a little more cumbersome because you have to worry about special cases.

What I would suggest is the following usage:
```bash
./replace_parameter.sh [-f <FILENAME>] <REGISTER> <LINK> <VALUE>
```
If `-f` (optional) is provided, no one cares whether or not `<VALUE>` has been supplied, but if it's not, then `<VALUE>` is used.
Regardless, `<REGISTER>` and `<LINK>` are always in the same positions.

You then have to use the `shift` mechanism in `getopts` to always get the correct index, regardless of chosen usage.
This will significantly clean up the logic at the end here as you simply have to do something like:
```bash
    ...
    REGISTER=$1
    LINK=$2
    VALUE=$3

    if [ -z ${FILENAME} ]
    then
        if [ -z ${VALUE} ]
        then
            echo 'No value supplied'
            usage
        else
            sed -i "s|^${REGISTER}.*|${REGISTER}   ${VALUE}|g" /mnt/persistent/gemdaq/vfat3/config_OH${LINK}_VFAT*_cal.txt
        fi
    else
        if [ -f ${FILENAME} ]
        then
            awk '{system("sed -i \"s|^'$REGISTER'.*|'$REGISTER'   "$2"|g\" /mnt/persistent/gemdaq/vfat3/config_OH'$LINK'_VFAT"$1"_cal.txt")}' $FILENAME
        else
            echo "File ${FILENAME} not found"
        fi
    fi

```

AndrewLevin commented:
@jsturdy, I exchanged the order of the <VALUE> and <LINK> arguments. I did it without using `shift`. Does it look OK?

bdorney commented:
Before this gets merged in, and especially after all these changes, @AndrewLevin could you attach examples showing that this is fully tested and behaves as expected for all use cases?

jsturdy commented:
Use `shift`; this may work now, but if you ever want to change it again to add different functionality, you'll have to add more and more special handling cases.
I suspect `shift` was put into the `getopt` library exactly to handle the case of positional arguments always being treated the same way

AndrewLevin commented:
I use `shift` now. I will proceed to testing.

jsturdy commented on scripts/replace_parameter.sh:
```diff
@@ -1,27 +1,69 @@
-
-	kill -INT $$; 
+	echo '		./replace_parameter.sh PULSE_STRETCH 1 4'
+	echo '		./replace_parameter.sh -f /path/to/NominalDacValues.txt PULSE_STRETCH 1'
```
why did you remove the `kill -INT $$`?
it would make all the `exit`s you have redundant (and sometimes `exit` can have unintended consequences, referring to @bdorney's comment about testing good and bad usage)

AndrewLevin replied:
I removed that `kill -INT $$` because @bdorney requested an `echo` statement after the usage: https://github.com/cms-gem-daq-project/gemctp7user/pull/38#discussion_r229235451. Now that `echo` statement comes before the usage, so I just added the `kill -INT $$` back.

bdorney replied:
My comment doesn't prevent you from using `kill -INT $$`.  The point is that the user should know what they did wrong.  The way you tell them this depends on the way they called the function.  Ideally you should:

1. provide the general use case,
2. if the script thinks they are trying to write all values to all registers and they screw up, provide that usage case,
3. if the script thinks they are providing a file to write unique values, provide that usage case.

This could be accomplished by having a general, and two specific `usage()` functions and still make use of `kill -INT $$`.

AndrewLevin replied:
OK, agreed. But I think the currently committed version with `kill -INT $$` in the `usage`  function and the `echo` statements on lines `45` and `53` resolves this issue, right?

AndrewLevin commented:
@jsturdy, I have updated the help information and fixed the indentation issues.

Also, I have performed the following tests:

```
eagle64:~$ cat /mnt/persistent/gemdaq/vfat3/config_OH0_VFAT*_cal.txt | grep BIAS_PRE_VREF
BIAS_PRE_VREF   89
BIAS_PRE_VREF   89
BIAS_PRE_VREF   95
BIAS_PRE_VREF   94
BIAS_PRE_VREF   91
BIAS_PRE_VREF   85
BIAS_PRE_VREF   95
BIAS_PRE_VREF   87
BIAS_PRE_VREF   94
BIAS_PRE_VREF   92
BIAS_PRE_VREF   96
BIAS_PRE_VREF   92
BIAS_PRE_VREF   90
BIAS_PRE_VREF   90
BIAS_PRE_VREF   92
BIAS_PRE_VREF   89
BIAS_PRE_VREF   88
BIAS_PRE_VREF   100
BIAS_PRE_VREF   87
BIAS_PRE_VREF   109
BIAS_PRE_VREF   91
BIAS_PRE_VREF   80
BIAS_PRE_VREF   88
BIAS_PRE_VREF   88
eagle64:~$ cat /mnt/persistent/gemdaq/vfat3/config_OH0_VFAT0_cal.txt                                                             
dacName/C:dacVal/I
PULSE_STRETCH           7
SYNC_LEVEL_MODE         0
SELF_TRIGGER_MODE       0
DDR_TRIGGER_MODE        0
SPZS_SUMMARY_ONLY       0
SPZS_MAX_PARTITIONS     0
SPZS_ENABLE             0
SZP_ENABLE              0
SZD_ENABLE              0
TIME_TAG                0
EC_BYTES                0
BC_BYTES                0
FP_FE                   7
RES_PRE                 1
CAP_PRE                 0
PT                     15
EN_HYST   1
SEL_POL                 1
FORCE_EN_ZCC            0
FORCE_TH                0
SEL_COMP_MODE           1
VREF_ADC                3
MON_GAIN                0
MONITOR_SELECT          0
IREF   32
THR_ZCC_DAC            10
THR_ARM_DAC           200
HYST   3
LATENCY                45
CAL_SEL_POL             1
CAL_PHI                 0
CAL_EXT                 0
CAL_DAC               200
CAL_MODE                1
CAL_FS                  0
CAL_DUR               200
BIAS_CFD_DAC_2   41
BIAS_CFD_DAC_1   41
BIAS_PRE_I_BSF         13
BIAS_PRE_I_BIT   214
BIAS_PRE_I_BLCC   22
BIAS_PRE_VREF   89
BIAS_SH_I_BFCAS   122
BIAS_SH_I_BDIFF   92
BIAS_SH_I_BFAMP         0
BIAS_SD_I_BDIFF   145
BIAS_SD_I_BSF   14
BIAS_SD_I_BFCAS   121
eagle64:~$ cat ~/NominalDACValues_GE11-X-S-INDIA-0002/2018.10.31.14.27/NominalDACValues.txt                                      
0       3
1       3
2       0
3       1
4       0
5       0
6       1
7       0
8       1
9       2
10      2
11      7
12      3
13      0
14      0
15      0
16      1
17      5
18      4
19      4
20      0
21      0
22      0
23      0
eagle64:~$ ./replace_parameter.sh -f ~/NominalDACValues_GE11-X-S-INDIA-0002/2018.10.31.14.27/NominalDACValues.txt BIAS_PRE_VREF 0
eagle64:~$ cat /mnt/persistent/gemdaq/vfat3/config_OH0_VFAT*_cal.txt | grep BIAS_PRE_VREF                                        
BIAS_PRE_VREF   3
BIAS_PRE_VREF   2
BIAS_PRE_VREF   7
BIAS_PRE_VREF   3
BIAS_PRE_VREF   0
BIAS_PRE_VREF   0
BIAS_PRE_VREF   0
BIAS_PRE_VREF   1
BIAS_PRE_VREF   5
BIAS_PRE_VREF   4
BIAS_PRE_VREF   4
BIAS_PRE_VREF   3
BIAS_PRE_VREF   0
BIAS_PRE_VREF   0
BIAS_PRE_VREF   0
BIAS_PRE_VREF   0
BIAS_PRE_VREF   0
BIAS_PRE_VREF   1
BIAS_PRE_VREF   0
BIAS_PRE_VREF   0
BIAS_PRE_VREF   1
BIAS_PRE_VREF   0
BIAS_PRE_VREF   1
BIAS_PRE_VREF   2
eagle64:~$ ./replace_parameter.sh BIAS_PRE_VREF 0 15                                                                             
eagle64:~$ cat /mnt/persistent/gemdaq/vfat3/config_OH0_VFAT*_cal.txt | grep BIAS_PRE_VREF
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
BIAS_PRE_VREF   15
eagle64:~$ cat /mnt/persistent/gemdaq/vfat3/config_OH0_VFAT0_cal.txt                     
dacName/C:dacVal/I
PULSE_STRETCH           7
SYNC_LEVEL_MODE         0
SELF_TRIGGER_MODE       0
DDR_TRIGGER_MODE        0
SPZS_SUMMARY_ONLY       0
SPZS_MAX_PARTITIONS     0
SPZS_ENABLE             0
SZP_ENABLE              0
SZD_ENABLE              0
TIME_TAG                0
EC_BYTES                0
BC_BYTES                0
FP_FE                   7
RES_PRE                 1
CAP_PRE                 0
PT                     15
EN_HYST   1
SEL_POL                 1
FORCE_EN_ZCC            0
FORCE_TH                0
SEL_COMP_MODE           1
VREF_ADC                3
MON_GAIN                0
MONITOR_SELECT          0
IREF   32
THR_ZCC_DAC            10
THR_ARM_DAC           200
HYST   3
LATENCY                45
CAL_SEL_POL             1
CAL_PHI                 0
CAL_EXT                 0
CAL_DAC               200
CAL_MODE                1
CAL_FS                  0
CAL_DUR               200
BIAS_CFD_DAC_2   41
BIAS_CFD_DAC_1   41
BIAS_PRE_I_BSF         13
BIAS_PRE_I_BIT   214
BIAS_PRE_I_BLCC   22
BIAS_PRE_VREF   15
BIAS_SH_I_BFCAS   122
BIAS_SH_I_BDIFF   92
BIAS_SH_I_BFAMP         0
BIAS_SD_I_BDIFF   145
BIAS_SD_I_BSF   14
BIAS_SD_I_BFCAS   121
eagle64:~$ 
```
